underscore.js

```diff
@@ -371,7 +363,19 @@
       if (!isArrayLike(obj)) obj = _.values(obj);
       return obj[_.random(obj.length - 1)];
     }
-    return _.shuffle(obj).slice(0, Math.max(0, n));
+
+    var set = isArrayLike(obj) ? obj : _.values(obj);
+    var length = getLength(set);
+    n = Math.min(n, getLength(set));
+    var sample = _.clone(set);
```

Comment:
One last nitpick: I think we can use a [modified version](http://en.wikipedia.org/wiki/Reservoir_sampling#Relation_to_Fisher-Yates_shuffle) of Fisher-Yates to only allocate an array of `n`.

```javascript
    var set = isArrayLike(obj) ? obj : _.values(obj);
    var length = getLength(set);
    n = Math.min(n, length);
    var sample = Array(n);
    var rand;
    for (var index = 0; index < n; index++) {
      rand = _.random(0, index);
      if (rand !== index) sample[index] = sample[rand];
      sample[rand] = set[index];
    }
    for (; index < length; index++) {
      rand = _.random(0, index);
      if (rand < n) sample[rand] = set[index];
    }
    return sample;
```